sys-utils/libcmt/include/libcmt/io.h

```diff
@@ -0,0 +1,165 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/** @file
+ * @defgroup libcmt_io_driver Cartesi Machine Input/Output Driver
+ * Low level abstraction of the Cartesi Machine kernel driver.
+ *
+ * Interaction is as follows:
+ * - Open the kernel device and mmap the shared memory ranges (tx and rx) with:
+ *   @ref cmt_io_init.
+ * - Retrieve the memory regions from the library with: @ref cmt_io_get_tx and
+ *   @ref cmt_io_get_rx.
+ * - Do requests (next input, emit output ...) by yielding the control back to
+ *   the emulator with: @ref cmt_io_yield.
+ *
+ * Lets look at some code:
+ *
+ * @include examples/io.c
+ *
+ * There are two implementations provided: `ioctl` that interacts with the
+ * cartesi-machine linux driver and `mock` that reads and writes the host
+ * filesystem, used for testing.
+ *
+ * @section Environment variables
+ *
+ * this module exposes two environment variables: @p CMT_DEBUG and @p CMT_INPUTS.
+ *
+ * @p CMT_DEBUG prints runtime information during application execution.
+ *
+ * ```
+ * CMT_DEBUG=yes ./application
+ * ```
+ *
+ * @p CMT_INPUTS feeds inputs to the mock (ignored for ioctl).
+ *
+ * ```
+ * CMT_INPUTS=0:advance.bin,1:inspect.bin ./application
+ * ```
+ *
+ * @ingroup libcmt
+ * @{ */
+#ifndef CMT_IO_H
+#define CMT_IO_H
+#include <stddef.h>
+#include <stdint.h>
+#include "buf.h"
+
+enum {
+	CMT_IO_DEV                 = 2, /**< IO Device */
+};
+
+/** Request */
+enum {
+	CMT_IO_CMD_AUTOMATIC       = 0, /**< Automatic */
+	CMT_IO_CMD_MANUAL          = 1, /**< Manual */
+};
+
+/** Request */
+enum {
+	CMT_IO_REASON_PROGRESS     = 0, /**< Progress */
+	CMT_IO_REASON_RX_ACCEPTED  = 1, /**< Accept and load next input */
+	CMT_IO_REASON_RX_REJECTED  = 2, /**< Reject and revert */
+	CMT_IO_REASON_TX_OUTPUT    = 3, /**< emit an output */
+	CMT_IO_REASON_TX_REPORT    = 4, /**< emit a report */
+	CMT_IO_REASON_TX_EXCEPTION = 5, /**< emit a exception and halt execution */
+};
+
+/** Reply reason when requesting @ref CMT_IO_REASON_RX_ACCEPTED */
+enum {
+	CMT_IO_REASON_ADVANCE      = 0, /**< Input is advance */
+	CMT_IO_REASON_INSPECT      = 1, /**< Input is inspect */
+};
+
+typedef struct cmt_io_yield cmt_io_yield_t;
+typedef union cmt_io_driver cmt_io_driver_t;
+
+struct cmt_io_driver_ioctl {
+	cmt_buf_t tx[1], rx[1];
+	int fd;
+};
+
+struct cmt_io_driver_mock {
+	cmt_buf_t tx[1], rx[1];
+	cmt_buf_t inputs_left;
+	int input_type;
+	char input_filename[128],
+	     input_fileext [16];
+
+	int input_seq,
+	    output_seq,
+	    report_seq,
+	    exception_seq;
```

Comment:
It is better to break each declaration into a line (make easier parsing in C binding generators). Eg:
```
int input_seq;
int output_seq;
```